### PR TITLE
IA-1832 Update GoogleDataprocService

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-1832

**Reason for version bump:**
Updated pubsub version (didn't realize we haven't updated it for a while), which deprecated `ProjectTopicName` in most APIs, hence replace `ProjectTopicName` with `TopicName` in our lib as well.

**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

### CODE REVIEW

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala [hunk 1]
```diff
@@ -125,6 +127,44 @@ private[google2] class GoogleDataprocInterpreter[F[_]: Async: StructuredLogger:
...
+object GoogleDataprocInterpreter {
+  def getAllInstanceNames(cluster: Cluster): Map[DataprocRole, Set[InstanceName]] = {
+    def getFromGroup(role: DataprocRole)(group: InstanceGroupConfig): Map[DataprocRole, Set[InstanceName]] =
+      group.getInstanceNamesList.asByteStringList().asScala.toList.map {
+        byteString =>
+          role -> Set(InstanceName(byteString.toStringUtf8))
+      }.toMap
+
+    val res = Option(cluster.getConfig).flatMap { config =>
+      val masters = Option(config.getMasterConfig).map(getFromGroup(DataprocRole.Master))
```
Comment (akarukappadath) [inline google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala]: there can only be one `master` right?

**Review by akarukappadath — APPROVED** [review]
lgtm! just a comment about a variable name

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala (hunk 1 above)
Comment (Qi77Qi) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala]: yea I think so...this code is copied from leonardo's existing code

Comment (Qi77Qi) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala]: oh I see...plural...wll fix!

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala [hunk 2]
```diff
@@ -125,6 +127,44 @@ private[google2] class GoogleDataprocInterpreter[F[_]: Async: StructuredLogger:
...
+      val workers = Option(config.getWorkerConfig).map(getFromGroup(DataprocRole.Worker))
+      val secondaryWorkers = Option(config.getSecondaryWorkerConfig).map(getFromGroup(DataprocRole.SecondaryWorker))
+
+      masters |+| workers |+| secondaryWorkers
+    }
+
+    res.getOrElse(Map.empty)
+  }
+}
```
Comment (kyuksel) [inline google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocInterpreter.scala]: Nit: add newline

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala [hunk 3]
```diff
@@ -36,29 +36,38 @@ trait GoogleDataprocService[F[_]] {
...
   private def fromCredential[F[_]: StructuredLogger: Async: Timer: ContextShift](
     googleCredentials: GoogleCredentials,
     blocker: Blocker,
+    regionName: RegionName,
     blockerBound: Semaphore[F],
     retryConfig: RetryConfig
   ): Resource[F, GoogleDataprocService[F]] = {
     val settings = ClusterControllerSettings
       .newBuilder()
+      .setEndpoint(s"${regionName.value}-dataproc.googleapis.com:443")
```
Comment (kyuksel) [inline google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala:70]: Should `443` come from a config?

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala [hunk 4]
```diff
@@ -90,3 +99,14 @@ object DeleteClusterResponse {
   final case class Success(clusterOperationMetadata: ClusterOperationMetadata) extends DeleteClusterResponse
   case object NotFound extends DeleteClusterResponse
 }
+
+sealed trait DataprocRole extends Product with Serializable
+object DataprocRole {
+  case object Master extends DataprocRole
+  case object Worker extends DataprocRole
+  case object SecondaryWorker extends DataprocRole
```
Comment (kyuksel) [inline google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala]: Perhaps add a comment that `Secondary Worker` = `Preemptible Worker`? I had not heard of the former before  🤷‍♂️

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala [hunk 5]
```diff
@@ -90,3 +99,14 @@ object DeleteClusterResponse {
...
+
+  val stringToDataprocRole = sealerate.values[DataprocRole].map(p => (p.toString, p)).toMap
+}
+
+final case class ClusterError(code: Int, message: String)
```
Comment (kyuksel) [inline google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala]: Nit: add newline (unless `scalafmt` does it for us)

**Review by kyuksel — APPROVED** [review]
👍 with minor comments inline

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala (hunk 3 above)
Comment (Qi77Qi) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala:70]: I don't think this is configurable...see https://cloud.google.com/dataproc/docs/concepts/regional-endpoints#using_regional_endpoints

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala (hunk 4 above)
Comment (Qi77Qi) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala]: This is straight copy from leonardo..I don't know if `Secondary Worker = Preemptible Worker` is true...but if you're sure that's the case, I can add the comment.

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala (hunk 3 above)
Comment (kyuksel) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala:70]: yea, looks like it

File: google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala (hunk 4 above)
Comment (kyuksel) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala]: > Preemptible workers are "secondary" workers; unlike "primary" workers, secondary workers do not store data. Therefore, you can use secondary workers to scale compute without scaling storage.

https://cloud.google.com/dataproc/docs/concepts/compute/preemptible-vms

Comment (Qi77Qi) [reply google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala]: 👍  I checked leo code as well...it does seem like leo considers secondary worker as preemptible..so added comment

**Pull request merged**